### PR TITLE
[SYCL] Remove memory allocation/free nodes

### PR DESCRIPTION
**Pull request by EwanC**
There are [concerns with the currently specified approach](https://github.com/intel/llvm/pull/5626#discussion_r1084170330) so remove the API and advise users to allocate memory before submitting the graph. Marked as draft for now as this requires meeting discussion.

Conflicts with semantics of these PRs, which should be closed if we merge this:
- https://github.com/reble/llvm/pull/66
- https://github.com/reble/llvm/pull/62

Addressed https://github.com/reble/llvm/issues/39

### CODE REVIEW

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc [hunk 1]
```diff
@@ -1118,6 +1048,45 @@ cgh.host_task([=](){
...
+non-performant to allocate or free memory as a node executed during graph
+submission. Instead such a memory allocation API needs to provide a way to
+return a pointer which won't be valid until the allocation is made on graph
+finalization, as allocating at finalization is the only way to benefit from
+the known graph scope for optimal memory allocation, and even optimize to
+eliminate some allocations entirely.
+
+Such a deferred allocation strategy presents challenges however, and as a result
+we recommend instead that prior to graph construction users perform core SYCL
+USM allocations to be used in the graph submission. Before to coming to this
```
Comment (Bensuo) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc:1087]:
```suggestion
USM allocations to be used in the graph submission. Before coming to this
```

**Review by Bensuo — APPROVED** [review]
LGTM, just a minor comment.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc [hunk 2]
```diff
@@ -1205,16 +1167,15 @@ int main() {
...
 
-  auto node_fx = g.add_free(x, {sycl_ext::property::node::depends_on(node_c)});
-  auto node_fy = g.add_free(y, {sycl_ext::property::node::depends_on(node_b)});
-
   auto exec = g.finalize(q.get_context());
 
   // use queue shortcut for graph submission
   q.ext_oneapi_graph(exec).wait();
 
   // memory can be freed inside or outside the graph
```
Comment (reble) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc:1197]:
```suggestion
  // memory is freed outside the graph
```

**Comment by sommerlukas** [comment]
As already discussed a bit over in the [graph fusion proposal](https://github.com/sommerlukas/llvm/pull/3), the current proposal for graph fusion builds on top of `add_malloc_device` to specify internalization of pointers through a property. 

Removing the `add_malloc_device` API from the graph proposal does not make internalization of USM pointers impossible, as the same property could be passed to the core USM functions `malloc_device/malloc_shared`, to mark a USM pointer for internalization (although this might slightly complicate the implementation). 

However, attaching the property to `add_malloc_device` has two advantages over attaching it to `malloc_device/malloc_shared`:
* Granularity: When attaching the property to `add_malloc_device`, it is possible to limit the scope of the internalization semantics to just the graph. When attaching the same property to `malloc_device/malloc_shared`, the internalization semantics would apply to *all* graphs using this USM pointer. Of course, this can be avoided by allocating device memory just for the use in the graph by the user, but users would need to make sure to not accidentally use a pointer with internalization semantics in a graph where this is not intended.
* Optimization: When the internalization property is attached to `add_malloc_device` and the JIT compiler is successful in his attempt to internalize during `graph::finalize`, it is possible to completely remove the device allocation from the graph. This can reduce the number of device allocations, potentially leading to higher performance. If the USM allocation happens outside of the graph and the graph has no control over it, this optimization won't be possible. 

tl;dr: Removing `add_malloc_device` will not make USM internalization in graph fusion impossible, but it may hinder some optimizations and require extra care from the user.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc [hunk 3]
```diff
@@ -1118,6 +1048,45 @@ cgh.host_task([=](){
...
+alive so that the reference is valid, which is unintuitive and is likely to
+result in bugs.
+
+Design 2) introduces a handle object which has the advantages of being a less
+error prone way to provide the pointer to the deferred allocation. However, it
+requires kernel changes and introduces an overhead above the raw pointers that
+are the advantage of USM.
+
+The backend support for design 3) is too limited for the approach to be
+feasible.
```
Comment (reble) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
Design 3) needs specific backend support for deferred allocation.
```

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc [hunk 4]
```diff
@@ -1118,6 +1048,45 @@ cgh.host_task([=](){
 });
 ----
 
+=== Memory Allocation Nodes
+
+There is no provided interface for users to define a USM allocation/free
+operation belonging to the scope of the graph. It would be error prone and
+non-performant to allocate or free memory as a node executed during graph
+submission. Instead such a memory allocation API needs to provide a way to
```
Comment (reble) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
submission. Instead, such a memory allocation API needs to provide a way to
```

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc [hunk 5]
```diff
@@ -1118,6 +1048,45 @@ cgh.host_task([=](){
...
+the known graph scope for optimal memory allocation, and even optimize to
+eliminate some allocations entirely.
+
+Such a deferred allocation strategy presents challenges however, and as a result
+we recommend instead that prior to graph construction users perform core SYCL
+USM allocations to be used in the graph submission. Before coming to this
+recommendation we considered the following explicit graph building interfaces
+for adding a memory allocation owned by the graph:
+
+1. Allocation function returning a reference to the raw pointer, i.e `void*&`,
```
Comment (reble) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
1. Allocation function returning a reference to the raw pointer, i.e. `void*&`,
```

**Review by reble — APPROVED** [review]
LGTM. Some minor comments.

**Comment by EwanC** [comment]
Merging this, since although we're in the middle of considering other ideas of the memory allocation API, the existing API isn't one we currently want to promote, so removing it better represents the current status of the feature.

Once we have a concrete idea of the memory allocation API we want to advocate, then we can create a PR for that. Even reverting this commit from the git history, in the case it does turn out we want to bring back what we had.
